README.md
# Binary Releases

If you don't want to build the Library or Editor to use it,
I release the latest binaries for the project in each GitHub release.

https://github.com/Teriks/LibLSLCC/releases


Each release contains an AnyCPU Release/Debug version of LibLSLCC,
an AnyCPU binary copy of lslcc (The command line compiler), and an XML documentation 
file generated for the library by Visual Studio/MSBuild.

Releases also include platform specific installers for LSLCCEditor.

However, the only difference between each installer is the install directory that they default too.
The installers for both x64 and x86 platforms use platform independent AnyCPU binaries.


# About LibLSLCC 


LibLSLCC is a framework for writing LSL compilers.

The LibLSLCC code validator/syntax tree builder provides full front end syntax checking of LSL.
It even includes extended warnings that you would find in most compilers for other languages,
but that are not implemented in either the Linden compiler or current OpenSim compiler.


Warnings for things such as: 

 * Constant expressions in conditional/loop statements.

 * Un-used variables.

 * Dead code.

 * Deprecated function usage.


That's not everything, just some common ones; the list is pretty long.
I suggest you mess around with the editor releases to discover what all 
LibLSLCC can tell you about your code.


LibLSLCC can also be used for general purpose LSL parsing tasks, it provides
its own rich syntax tree that has been completely abstracted from ANTLR.

The syntax tree is tailored specifically for dealing with LSL, and there is an interface for every node 
so that you can implement your own code DOM to feed to compilers/formatters if you want to.


====


LibLSLCC includes a CSharp code generator that targets the OpenSim runtime.

The code validator (parser/tree builder) and OpenSim code generator in LibLSLCC have both been designed
with the intent of implementing Linden LSL with near perfect cross compatibility.

The project is basically a complete reverse engineering of the Linden compiler's
grammar, non grammar level syntax rules, and generated code behavior.



I have integrated LibLSLCC into OpenSim, see here:

https://github.com/Teriks/OpenSim_With_LibLSLCC


====

A few notable code validator/OpenSim code generator features:

 * Full front end Syntax Checking, including dead code detection.  
   Esoteric CSharp compiler errors and line mapping have been eliminated.

 * Dead code elimination from generated code where applicable.
   This includes un-used functions and global variables, as well
   as any dead code in a function/event body that does not cause a compile 
   error and is safe to remove given its context.

 * Correct code generation for global variables that reference each other
   while still maintaining full support for script resets in server generated code.

 * Symbol name mangling specific to globals/parameters/local variables/labels and 
   user defined functions.  This completely abstracts variable scoping rules from the CSharp 
   compiler underneath.  All variable scoping rules are handled by the front end LibLSLCC code validator
   and are implemented in a fashion that is true to Linden LSL.  Symbol mangling also removes the possibility 
   of causing a CSharp syntax error by using a keyword/Class name as a variable or function name.


 * Correct Code generation for jumps over variable declarations.  Jumping over a variable declaration
   will leave said variable with a default value instead of null.  Leaving these variables null is 
   something both the Linden compiler and current OpenSim compiler do, which is a step out of the LSL
   SandBox, as null variables are not supported elsewhere in LSL and generally not expected as a function parameter.

 * Labels no longer require a "NoOp()" after every label, the OpenSim code generator 
   in LibLSLCC simply detects when they are the last statement in a scope, then adds a semi-colon
   immediately after them in the generated code.

 * Full and optimized support for the CO-OP script stop strategy in OpenSim.

 * Correct order of operations via the use of operator function stubs
   that are generated on demand.  Old mono list optimizations will now
   port over without breaking, as well as other odd scripts that rely
   on Right to Left evaluation being the norm.


 * Negatable vectors and rotations.

 * Support for certain unicode characters in script symbols. (Same characters the linden Compiler allows)



# About LSLCCEditor

The project also includes an LSL Editor that was originally built to test the compiler library.
It has since developed into a full featured multi-tabbed IDE that is built on top of LibLSLCC's parsing framework.

Please take note that LSLCCEditor and its related projects are **Windows Only**, and you can 
only build the __LibLSLCC-WithEditor*.sln__ solution's on Windows using a version of Visual Studio that
support's targeting the .NET 4.5 framework.



VS2015+ is your best bet for building, however if your version of Visual Studio's
support's .NET 4.5 you will probably be able to build the Editor project files with it.


LSLCCEditor features:

 * Tabbed multi script editing.

 * Syntax highlighting.

 * Documentation tooltips.

 * Context aware auto complete.

 * Go to definition (navigation by symbol).

 * Code formatting.

 * Library data for both Linden and Opensim SecondLife servers. 

 * Compilation to CSharp code for OpenSim as either
    client uploadable code or plain server side code. 

===

A few images of the editor small enough to squeeze into this readme...  

This dark color scheme can be found here: https://github.com/Teriks/LSLCCEditorThemes


![LSLCCEditor](http://imgur.com/qOoVhjK.png)

![LSLCCEditor](http://imgur.com/hwIOHAf.png)


# Project Dependencies

All binary dependencies are distributed with the build.
You should not need to add any binaries yourself.

===

The LibLSLCC library itself depends on the **CSharp ANTLR 4 Runtime**.

You can find a license for ANTLR 4 in the **ThirdPartyLicenses** folder
of the build directory.

For more information about **ANTLR** go here:

http://www.antlr.org/index.html


===

The **LibraryDataScrapingTool** project depends on **System.Data.SQLite**
when compiled on Windows.  When its compiled on mono it uses **Mono.Data.Sqlite**.

see here: 

https://system.data.sqlite.org/index.html/doc/trunk/www/index.wiki


===

The LSLCCEditor application depends on **AvalonEdit**, and a few pieces
of code forked from **AvalonEdit**.  You can also find a license for **AvalonEdit**
in the **ThirdPartyLicenses** folder of the build directory.

 
For more information about **AvalonEdit** go here:

http://avalonedit.net/


=== 


 
# Building LibLSLCC and LSLCCEditor On Windows

Important thing to note:

 * Building **LibLSLCC-WithEditor-WithInstaller.sln** should work with **VS2015 and up**.

 * Building **LibLSLCC-WithEditor-NoInstaller.sln** should work with **VS2012 and up**.

 * Building **LibLSLCC-NoEditor.sln** should work with **VS2010 and up**.


===


LibLSLCC requires Java to be installed so that the ANTLR 4 parser generator tool 
can be run as part of the pre-build step for the LibLSLCC Library.  

The java executable will need to be in your PATH so that it can run from the command prompt. 
Fortunately installing java itself sets this up for you by default.


In order to build the **LibLSLCC-WithEditor-WithInstaller.sln** solution on Windows you must 
install the latest WiX Installer Toolset from http://wixtoolset.org/releases/.

You must also install the WiX Toolset Extension for your version of Visual Studio, which can
also be found at http://wixtoolset.org/releases/.

Installing the extension will allow you to load the WiX Project type, so you can build the 
**LSLCCEditorInstaller** project that is part of the **LibLSLCC-WithEditor-WithInstaller.sln** solution file.


**LibLSLCC-WithEditor-WithInstaller.sln** will only build the installer project when building for
**x64** or **x86** in **Release** mode.  The WiX installer build will not be triggered if you have the
"Platform" setting set to "AnyCPU" in Visual Studio, you must change it to either **x64** or **x86**.


The **LSLCCEditor** related projects are only buildable on Windows as they depend on WPF and AvalonEdit, 
but the **LibLSLCC** library, tools, and command line compiler are cross platform.  


Additionally the **LSLCCEditor** related projects can only be built in versions of Visual Studio 
that support targeting the .NET 4.5 framework. (VS2012+)


I have not tested other versions of Visual Studio besides VS2010 with the **LSLCCEditor** portion of the build.
VS2010 for sure **does not work**, as the editor uses a .NET framework level (4.5) that is too high for it to target.

If you happen to be using a version of Visual Studio that is incompatible with the **LSLCCEditor** portion of the
build, and you just want to build the library portion of the project;  Then you should use **LibLSLCC-NoEditor.sln**.


It contains only the library project and projects from the source tree which have a .NET 4.0 compatibility profile.
The projects in the **LibLSLCC-NoEditor.sln** solution have been tested with VS2010 so far, they will most likely
build with every version of Visual Studio starting with VS2010.


# Building LibLSLCC on *Nix Platforms with mono/xbuild


To build LibLSLCC from the command line using mono/xbuild on nix platforms, first install Mono (complete) 
for your distribution and then CD into the LibLSLCC source tree where **LibLSLCC-NoEditor.sln** resides.

enter the command:

	xbuild /p:Configuration=Release LibLSLCC-NoEditor.sln


LibLSLCC should start building.


**Note about xbuild when using Mono 4.* on *Nix:**


The latest versions of mono starting with major version 4 cannot build assemblies targeting .NET Framework v4.0

If your building using 4.* version of mono you need to force all projects to target the v4.5 framework, with:

	xbuild /p:Configuration=Release /p:TargetFrameworkVersion="v4.5" LibLSLCC-NoEditor.sln


The python build script will do that for you behind the scenes.


# Cross Platform Python Build Scripts

The Python build scripts require python 3.* and the pip package manager for python3 to be installed.  You must also have an appropriate version of 
Visual Studios+WiX installed on Windows, or the Mono (complete) development package if your on *nix.

pip is installed with Python on Windows by default when using the offical installer package, on Linux or other platforms you may need to install it manually.  If you do not have pip (As indicated by the error produced when running build.py for the first time without pip installed), you should download the pip bootstrap script from <https://bootstrap.pypa.io/get-pip.py> and run it with python3 under admin or super user privileges.  If your distro has a python3-pip package already, you can also install using your package manager.

If your using Windows, you must install python and add it to your path if you want to call this script from the command line (There is an option to do this automatically from the installer dialog).
When installing python on Windows there is an option to associate .py files with the python executable, so that clicking on .py
files inside the file explorer will run them as well.

**build.py** can be used to build the library and optionally create a timestamped binary release zip file.

When build.py first runs it will download any dependencies it needs to the 'BuildScriptLibs' folder in the project tree, after that is done it will continue doing anythin you requested it to do.

Simply running `python build.py` will build Release and Debug versions of all projects that are buildable on your platform.

Use this to display additional build options:
 
	python build.py --help

The quickest way to build all the Release mode binaries for each project that will build on your platform is:

	python build.py --only-release

And to build just LibLSLCC:

	python build.py --only-liblslcc

To build and package up a version stamped release in the BinaryRelease folder:

	python build.py --binary-release

To clean the entire build:

	python build.py --clean

To update assembly versions (according to version.json):

	python build.py --update-versions


build.py --update-versions is meant to be used as a pre-commit hook.


CHANGELOG.md
# (1/26/2016 ...)<br/> Event handler cleanup / URL colors

Cleaned up event handler parameter names so that they formatted more traditionally when generating handlers using auto complete.

They were previously capitalized due to the source the library data was scraped from.

Changing the highlighting color of URL's in source code from the theme editor pane is now possible.

Its at: `Settings -> Editor Theme -> Syntax Highlighting / Editor Color -> Urls`


# (1/31/2016 ...)<br/> Fix to code formatter

Fixed an issue with formatting braceless conditional/loop expressions containing comments before the
single statement that makes up the code body.

Previously code such as this:


```LSL

if(TRUE)
    //test
    llOwnerSay("test");

```

Was being formatted like this, when the 'Indent Braceless Control Statements'
option was disabled in the formatter settings:


```LSL

if(TRUE) //test llOwnerSay("test");

```

This is of course a syntax error,  braceless expressions with any sort of comments
preceding the body expression are now forcefully indented regardless of the indent
settings for braceless control/loop statements.


# (1/31/2016 ...)<br/> Code generation change

Remove unnecessary code generation for implicit boolean conversions of vector/rotation/list and string.


# (2/4/2016 7:44AM)<br/> Fix improper code generation issue


Expression statements such as:

~(x=0);

-(x=0);

(unary op) (expression);


and:


(x=3);

(basically any expression in parenthesis as statement);


Were not being accounted for when compiling to CSharp.

CSharp does not allow anything but variable declarations,
variable assignments/modifications, and method calls to be used
as a standalone statement.  You cannot have a unary expression or 
parenthesized expression be a standalone statement.

Therefore these valid LSL expressions must be sanitized when compiling.

This was fixed simply by forcing them to become an argument to a generic
function that does nothing.

public static void ForceStatement<T>(T value){}


# (2/4/2016 11:44PM)<br/> Give logical AND and OR equal precedence

In Linden LSL `(0 && 0 || 1) == TRUE`, and `(1 || 0 && 0) == FALSE`

This seems to imply that whichever operator appears first has precedence.

&& and || have been given equal precedence to achieve this behavior.

===

Adjust code generation for boolean conversion of strings in condition statements

Added back `UTILITES.ToBool` to generated code for string expressions in conditional statements.
Because some string constants may be defined as a CSharp built in string instead of an OpenSim runtime string, which cannot implicitly convert to bool.


# (2/5/2016 11:58AM)<br/> Make modifiable LValues part of grammar

Having:

`expression '=' expression`

And:

`expression '*=' expression`

Etc..

In the expression grammar is causing strange parse trees.



Factoring it down to:

```

lvalue:  memberAccess | ID;

expression: ... expression stuff ...

    ...
    
    | lvalue ALL_MOD_ASSIGN_OPS expression
    | lvalue '=' expression
    
```

Helps monumentally...


This also happens to fix the precedence problem with expressions such as `-x = y`.


The downside to this is that assignments to non lvalues will
now result in a non-pretty grammar level syntax error.


The upside is that it fixes several operator precedence issues that
cause incompatibilities with SecondLife,  I currently do not know of any more (yet).

===

Reduced the size of generated OpenSim code.

The variable name mangling scheme has been changed so that symbol names are shorter.

Binary operator stub's also have a new naming scheme.

===

Fix issue with incorrectly detecting invalid escape codes in strings.

Behavior has been changed to simply replace the invalid escape code with the code 
character that comes after the slash, as in Linden LSL.


# (2/7/2016 4:26AM)<br/> Factor grammar to remove control statement ambiguity


A problem with the parser grammar for control statements was causing large performance issues when parsing long
if-elseif-else chains, sometimes to the point of putting the parser in an infinite loop.

This has been fixed by simplifying the grammar and refactoring the code validator class and autocomplete parser.

The abstracted tree LibLSLCC produces still retains the same structure.


# (2/8/2016 2:58AM)<br/> Added built in input file globbing to lslcc_cmd

Added built in file globbing and the ability to specify
multiple input files or glob expressions at once when invoking lslcc.exe.

Output file name templating, along with the capability to log
to a single file or multiple files (one per input file) has been added.

See lslcc.exe -h for details on how to use these.

This was added mostly for running the compiler over a ton of
files at once for testing purposes.

===

Also fixed an issue with LSLCCEditor's application settings.

Carriage return sequences in XML text content were being
converted into plain newlines upon saving the settings file.

This no longer happens, as carriage returns are now turned into
XML entities by the settings serializer when settings are persisted.

The application settings fix really only affected how script headers
under the compiler settings menu were formatted in generated code.


# (2/9/2016 4:38AM)<br/> Re-Implement internal LSLLabelCollectorPrePass

This class was not fixed properly after large changes to the ANTLR grammar.

I left it in a state that was breaking return path analysis pretty badly.  oops.

I commented out the visit hooks that were removed by the grammar changes and then forgot to fix it.

Return path analysis for jumps and labels was broken because this pre-pass visitor 
was not defining proper statement indices and code scope ID's for labels used inside of 
events and functions.

===

Fixed code generation for integer/hex literals that overflow/underflow LSL's integer type.

Fixed an issue with the code generation behavior for integer/hex literals that are too

build.py
#!/usr/bin/python3
import sys
import os

if sys.version_info[0] == 2:
    print('Please use python3 to run this script, preferably version 3.4+')
    exit()

script_path = os.path.dirname(os.path.realpath(__file__))
sys.path.insert(0, os.path.join(script_path, 'BuildScriptLibs'))


msbuildpy_version = '0.5.1.0'
msbuildpy_pip_install_target = 'git+https://github.com/Teriks/msbuildpy.git@'+msbuildpy_version


import zipfile
import datetime
import shutil
import platform
import subprocess
import re
import json
import importlib
import time
from argparse import ArgumentParser
from argparse import RawTextHelpFormatter


def set_version(file, version):
    with open(file, 'r') as content_file:
        content = content_file.read();

    match = re.search(r'\[assembly: AssemblyVersion\("([^"]*)"\)\]', content)

    was = match.group(1)

    if was == version:
        return was

    content = re.sub(r'\[assembly: AssemblyVersion\("[^"]*"\)\]',
                     r'[assembly: AssemblyVersion("' + version + '")]', content)

    content = re.sub(r'\[assembly: AssemblyFileVersion\("[^"]*"\)\]',
                     r'[assembly: AssemblyFileVersion("' + version + '")]', content)

    tempfile = file + ".tmp";

    with open(tempfile, 'w+') as content_file:
        content_file.write(content)

    # if theres an exception prior, the temp file will not move over.
    # an exception above will likely cause the temp file to be blank.
    # blanking the source file would be bad.
    shutil.move(tempfile, file)
    return was

    
def update_assembly_versions():
    print("")
    print("Updating assembly versions...")
    print("")

    scriptDir = os.path.dirname(os.path.realpath(__file__));

    versionsDir = os.path.join(scriptDir, "versions")

    with open("version.json", 'r') as content_file:
        versionFileContent = json.load(content_file)

    for dir in versionFileContent.keys():

        versionFile = os.path.join(versionsDir, dir, "Version.cs")

        lastTag = versionFileContent[dir]["last_tag"]

        commitsSinceLastTag = int(
            subprocess.check_output(["git", "rev-list", lastTag.rstrip() + "..HEAD", "--count"]).decode("utf-8"))

        versionTemplate = versionFileContent[dir]["version_template"]

        versionTemplate = versionTemplate.replace("{commits_since_last_tag}", str(commitsSinceLastTag))

        was = set_version(versionFile, versionTemplate)

        if versionTemplate != was:
            print(dir + " = " + versionTemplate + " was " + was)
        else:
            print(dir + " version already up to date.")


    print("")


def get_liblslcc_version():
    with open(os.path.join(script_path, "versions", "LibLSLCC", "Version.cs"), 'r') as content_file:
        content = content_file.read()

    match = re.search(r'\[assembly: AssemblyVersion\("([^"]*)"\)\]', content)
    return match.group(1)


def get_lslcceditor_version():
    with open(os.path.join(script_path, "versions", "LSLCCEditor", "Version.cs"), 'r') as content_file:
        content = content_file.read()

    match = re.search(r'\[assembly: AssemblyVersion\("([^"]*)"\)\]', content)
    return match.group(1)


def zip_dir_relative(path, zip_file, **kwargs):

    into_arch_dir = kwargs.get('archTopDir', None)
    rel_dir_transform = kwargs.get('archDirTransform', None)
    file_filter = kwargs.get('fileFilter', None)
    
    for root, dirs, files in os.walk(path):
        for file in files:
            full_path = os.path.join(root, file)
            rel_path = os.path.relpath(root, script_path)

            arc_path = os.path.join(rel_path, file)

            if rel_dir_transform is not None:
                arc_path = rel_dir_transform(arc_path)

            if into_arch_dir is not None:
                arc_path = os.path.join(into_arch_dir, rel_path, file)
                
            if file_filter is not None:
                if not file_filter(full_path):
                    continue

            print('Zip:\n\t' + os.path.relpath(full_path, script_path) + ' -> ' + arc_path)
            zip_file.write(full_path, arc_path)


def remove_second_folder_down(path):
    pathparts = path.split(os.sep)
    pathparts.pop(1)
    return os.path.join(*pathparts)


LibLSLCC_Version = get_liblslcc_version()

LSLCCEditor_Version = get_lslcceditor_version()

args_parser = ArgumentParser(formatter_class=RawTextHelpFormatter)

args_parser.add_argument(
    '--only-release',
    action='store_true',
    dest='release_only',
    help='Only build Release binaries.\n\n'
)

args_parser.add_argument(
    '--only-debug',
    action='store_true',
    dest='debug_only',
    help='Only build Debug binaries.\n\n'
)

args_parser.add_argument(
    '--only-liblslcc',
    action='store_true',
    dest='only_build_liblslcc',
    help='Only build the LibLSLCC library project and nothing else.\n\n'
)

args_parser.add_argument(
    '--liblslcc-net-v45',
    action='store_true',
    dest='liblslcc_net_45',
    help='Force the LibLSLCC library project to target .NET 4.5 instead of .NET 4.0.\n\n'
)

args_parser.add_argument(
    '--no-scraping-tool',
    action='store_false',
    dest='build_scraper',
    help='Prevent the LibraryDataScrapingTool project from being built.\n\n'
)

args_parser.add_argument(
    '--no-lslcc-cmd',
    action='store_false',
    dest='build_lslcc_cmd',
    help='Prevent the lslcc_cmd project from being built.\n\n'
)

args_parser.add_argument(
    '--no-demo-area',
    action='store_false',
    dest='build_demo_area',
    help='Prevent the DemoArea project from being built.\n\n'
)

args_parser.add_argument(
    '--no-editor',
    action='store_false',
    dest='build_editor',
    help='Prevent both the editor and installer from being built.\nOn Mono these do not build regardless.\n\n'
)

args_parser.add_argument(
    '--no-installer',
    action='store_false',
    dest='build_installer',
    help='Prevent the installer from being built if you do not have WiX.\nOn Mono it does not build regardless.\n\n'
)

args_parser.add_argument(
    '--binary-release',
    action='store_true',
    dest='make_binary_release',
    help=
    'Create a version stamped zip file of LibLSLCC and lslcc_cmd (plus installer file\'s on windows if --no-installer is not specified),\n' +
    'inside the directory specified by --binary-release-dir (Default directory is BinaryRelease).\n\n'
)

args_parser.add_argument(
    '--binary-release-dir',
    metavar='PATH',
    default=os.path.join(script_path, 'BinaryRelease'),
    dest='binary_release_dir',
    help='The folder to create and place the binary release files in.'
)

args_parser.add_argument(
    '--clean',
    action='store_true',
    dest='clean_build',
    help='Clean the build.'
)

args_parser.add_argument(
    '--update-versions',
    action='store_true',
    dest='update_versions',
    help='Update assembly versions in accordance with version.json.'
)


args = args_parser.parse_args()


if args.update_versions:
    update_assembly_versions()
    exit()



def install_deps():
    try:
        import pip
    except ImportError:
        print('Please install pip package manager for python3, see README.md for help')
        exit()
    pip.main(['install', '--upgrade', '--target', os.path.join(script_path, 'BuildScriptLibs'), msbuildpy_pip_install_target])    


def re_run():
    time.sleep(1)
    subprocess.call([sys.executable, os.path.realpath(__file__)]+sys.argv[1:])
    exit()


try:
    import msbuildpy
    if msbuildpy.__version__ != msbuildpy_version: 
        install_deps()
        re_run()
except ImportError:
    install_deps()
    re_run()


import msbuildpy.sysinspect


if msbuildpy.sysinspect.is_windows():
    MSBuild = msbuildpy.find_msbuild('msbuild >=12.*')
    if len(MSBuild) == 0:
        print('Could not find a compatible version of msbuild')
        
    # get the most recent major version
    MSBuild = MSBuild[0]
else:
    MSBuild = msbuildpy.find_msbuild('xbuild >=12.*')
    if len(MSBuild) == 0:
        print('Could not find a compatible version of xbuild')
        
    # get the most recent major version
    MSBuild = MSBuild[0]


def call_msbuild(*args):
    subprocess.call([MSBuild.path]+list(args))


if args.clean_build:
    if msbuildpy.sysinspect.is_windows():
        solution = os.path.join(script_path, 'LibLSLCC-WithEditor-WithInstaller.sln')
    else:
        solution = os.path.join(script_path, 'LibLSLCC-NoEditor.sln')


    call_msbuild(solution, '/t:clean', '/p:Configuration=Release','/p:Platform=Any CPU',)
    call_msbuild(solution, '/t:clean', '/p:Configuration=Debug','/p:Platform=Any CPU',)

    call_msbuild(solution, '/t:clean', '/p:Configuration=Release','/p:Platform=x86',)
    call_msbuild(solution, '/t:clean', '/p:Configuration=Debug','/p:Platform=x86',)

    call_msbuild(solution, '/t:clean', '/p:Configuration=Release','/p:Platform=x64',)
    call_msbuild(solution, '/t:clean', '/p:Configuration=Debug','/p:Platform=x64',)
    exit()


if args.only_build_liblslcc:
    args.build_scraper = False
    args.build_lslcc_cmd = False
    args.build_demo_area = False
    args.build_installer = False
    args.build_editor = False


if args.make_binary_release:
    args.build_scraper = False
    args.build_demo_area = False


no_editor_solution = os.path.join(script_path, 'LibLSLCC-NoEditor.sln')
editor_solution = os.path.join(script_path, 'LibLSLCC-WithEditor-WithInstaller.sln')

LSLCCEditorTargetFramework = "/p:TargetFrameworkVersion=4.5"
LibLSLCCTargetFramework = "/p:TargetFrameworkVersion=v4.0"

mono_vm = msbuildpy.sysinspect.get_mono_vm()


# mono 4.x / cannot build v4.0 assemblies
if args.liblslcc_net_45 or (not msbuildpy.sysinspect.is_windows() and mono_vm.version[0] > 3):
    LibLSLCCTargetFramework = "/p:TargetFrameworkVersion=v4.5"

LibLSLCCBuildTargets = "/t:LibLSLCC"

if args.build_lslcc_cmd:
    LibLSLCCBuildTargets += ";lslcc_cmd"


if args.build_scraper:
    LibLSLCCBuildTargets += ";LibraryDataScrapingTool"


if args.build_demo_area:
    LibLSLCCBuildTargets += ";DemoArea"


if not args.release_only:
    call_msbuild(no_editor_solution, LibLSLCCBuildTargets, '/p:Configuration=Debug', '/p:Platform=Any CPU',
                LibLSLCCTargetFramework)


if not args.debug_only:
    call_msbuild(no_editor_solution, LibLSLCCBuildTargets, '/p:Configuration=Release', '/p:Platform=Any CPU',
                LibLSLCCTargetFramework)


# build the installers on windows
if msbuildpy.sysinspect.is_windows() and args.build_installer and args.build_editor:
    if not args.release_only:
        call_msbuild(editor_solution, '/t:LSLCCEditor', '/p:Configuration=Debug', '/p:Platform=Any CPU',
                    LSLCCEditorTargetFramework)

    call_msbuild(editor_solution, '/t:LSLCCEditorInstaller', '/p:Configuration=Release', '/p:Platform=x86',
                '/p:Version=' + LSLCCEditor_Version, LSLCCEditorTargetFramework)
    call_msbuild(editor_solution, '/t:LSLCCEditorInstaller', '/p:Configuration=Release', '/p:Platform=x64',
                '/p:Version=' + LSLCCEditor_Version, LSLCCEditorTargetFramework)


if msbuildpy.sysinspect.is_windows() and not args.build_installer and args.build_editor:
    if not args.release_only:
        call_msbuild(editor_solution, '/t:LSLCCEditor', '/p:Configuration=Debug', '/p:Platform=Any CPU',
                    LSLCCEditorTargetFramework)

    if not args.debug_only:
        call_msbuild(editor_solution, '/t:LSLCCEditor', '/p:Configuration=Release', '/p:Platform=Any CPU',
                    LSLCCEditorTargetFramework)


if not args.make_binary_release:
    exit(0)

print('===========================================')
print('\n')

installerBasicName = 'LSLCCEditorInstaller'

installerExtension = '.msi'

binaryReleaseOutputDir = args.binary_release_dir

binariesZipPath = os.path.join(binaryReleaseOutputDir, 'LibLSLCC_Binaries_' + LibLSLCC_Version + '.zip')

LibLSLCC_Path = os.path.join(script_path, 'LibLSLCC')

lslcc_cmd_Path = os.path.join(script_path, 'lslcc_cmd')

installerPath = os.path.join(script_path, 'LSLCCEditorInstaller')

LibLSLCC_AnyCpu_Path = os.path.join(LibLSLCC_Path, 'bin', 'AnyCPU')

LibLSLCC_ThirdPartyLicenses_Path = os.path.join(LibLSLCC_Path, 'bin', 'ThirdPartyLicenses')

LibLSLCC_Licence_Path = os.path.join(LibLSLCC_Path, 'bin', 'LICENSE')

lslcc_cmd_AnyCpu_Path = os.path.join(lslcc_cmd_Path, 'bin', 'AnyCPU')

lslcc_cmd_ThirdPartyLicenses_Path = os.path.join(lslcc_cmd_Path, 'bin', 'ThirdPartyLicenses')

if not os.path.isdir(binaryReleaseOutputDir):
    os.mkdir(binaryReleaseOutputDir)

try:
    import zlib

    zipMode = zipfile.ZIP_DEFLATED
except:
    zipMode = zipfile.ZIP_STORED
    
    
def zip_extension_filter(path):
    return os.path.splitext(path)[1] != ".tmp"

# make the timestamped binary release zip file
with zipfile.ZipFile(binariesZipPath, 'w', zipMode) as zip_file:
    if not args.debug_only:
        zip_dir_relative(os.path.join(LibLSLCC_AnyCpu_Path, "Release"), zip_file,
                         archDirTransform=remove_second_folder_down,
                         fileFilter=zip_extension_filter)

    if not args.release_only:
        zip_dir_relative(os.path.join(LibLSLCC_AnyCpu_Path, "Debug"), zip_file,
                         archDirTransform=remove_second_folder_down,
                         fileFilter=zip_extension_filter)

    zip_dir_relative(LibLSLCC_ThirdPartyLicenses_Path, zip_file, 
                     archDirTransform=remove_second_folder_down, 
                     fileFilter=zip_extension_filter)

    zip_dir_relative(lslcc_cmd_ThirdPartyLicenses_Path, zip_file, 
                     archDirTransform=remove_second_folder_down, 
                     fileFilter=zip_extension_filter)

    lslcc_cmd_Arch_Path = os.path.basename(lslcc_cmd_Path)

    # only the release build of lslcc_cmd gets put in the zip
    if args.build_lslcc_cmd and not args.debug_only:
        zip_dir_relative(os.path.join(lslcc_cmd_AnyCpu_Path, "Release"), zip_file,
                         archDirTransform=remove_second_folder_down,
                         fileFilter=zip_extension_filter)

    zip_file.write(LibLSLCC_Licence_Path, os.path.basename('LICENSE'))

# copy and time stamp the installers when on windows
if msbuildpy.sysinspect.is_windows():
    x64_installer = os.path.relpath(
        os.path.join(installerPath, 'bin', 'x64', 'Release', installerBasicName + installerExtension), script_path)
    x86_installer = os.path.relpath(
        os.path.join(installerPath, 'bin', 'x86', 'Release', installerBasicName + installerExtension), script_path)

    x64_installerDest = os.path.relpath(
        os.path.join(binaryReleaseOutputDir, installerBasicName + '_x64_' + LSLCCEditor_Version + installerExtension),
        script_path)
    x86_installerDest = os.path.relpath(
        os.path.join(binaryReleaseOutputDir, installerBasicName + '_x86_' + LSLCCEditor_Version + installerExtension),
        script_path)

    if args.build_installer:
        print('\n')
        print('===========================================')
        print('\n')

        print('Copy x64 Installer:\n\t' + x64_installer + ' -> ' + x64_installerDest)
        shutil.copy2(x64_installer, x64_installerDest)

        print('\n')

        print('Copy x64 Installer:\n\t' + x86_installer + ' -> ' + x86_installerDest)
        shutil.copy2(x86_installer, x86_installerDest)

        print('\n')
